samples/java-eventsourced-shopping-cart/src/main/java/com/example/shoppingcart/domain/ShoppingCartImpl.java

```diff
@@ -16,94 +16,111 @@
+public class ShoppingCartImpl extends ShoppingCartInterface2 {
   @SuppressWarnings("unused")
   private final String entityId;
 
-  private final Map<String, ShoppingCartApi.LineItem> cart = new LinkedHashMap<>();
-
   public ShoppingCartImpl(@EntityId String entityId) {
     this.entityId = entityId;
   }
 
   @Override
-  public ShoppingCartDomain.Cart snapshot() {
-    return ShoppingCartDomain.Cart.newBuilder()
-        .addAllItems(cart.values().stream().map(this::convert).collect(Collectors.toList()))
-        .build();
+  protected ShoppingCartDomain.Cart emptyState() {
+    return ShoppingCartDomain.Cart.getDefaultInstance();
   }
 
   @Override
-  public void handleSnapshot(ShoppingCartDomain.Cart cart) {
-    this.cart.clear();
-    for (ShoppingCartDomain.LineItem item : cart.getItemsList()) {
-      this.cart.put(item.getProductId(), convert(item));
+  public Effect<Empty> addItem(
+          ShoppingCartApi.AddLineItem command,
+          ShoppingCartDomain.Cart currentState,
+          EventSourcedEntityEffect.Builder<Empty, ShoppingCartDomain.Cart> effectBuilder,
+          CommandContext ctx) {
+    if (command.getQuantity() <= 0) {
+      return effectBuilder.failure("Cannot add negative quantity of to item" + command.getProductId());
     }
-  }
 
-  @Override
-  protected Empty addItem(ShoppingCartApi.AddLineItem item, CommandContext ctx) {
-    if (item.getQuantity() <= 0) {
-      throw ctx.fail("Cannot add negative quantity of to item" + item.getProductId());
-    }
-    ctx.emit(
-        ShoppingCartDomain.ItemAdded.newBuilder()
+    ShoppingCartDomain.ItemAdded event = ShoppingCartDomain.ItemAdded.newBuilder()
             .setItem(
-                ShoppingCartDomain.LineItem.newBuilder()
-                    .setProductId(item.getProductId())
-                    .setName(item.getName())
-                    .setQuantity(item.getQuantity())
-                    .build())
-            .build());
-    return Empty.getDefaultInstance();
+                    ShoppingCartDomain.LineItem.newBuilder()
+                            .setProductId(command.getProductId())
+                            .setName(command.getName())
+                            .setQuantity(command.getQuantity())
+                            .build())
+            .build();
+
+    return effectBuilder
+            .emitEvent(event)
+            .thenReply(newState -> Empty.getDefaultInstance());
   }
 
   @Override
-  protected Empty removeItem(ShoppingCartApi.RemoveLineItem item, CommandContext ctx) {
-    if (!cart.containsKey(item.getProductId())) {
-      throw ctx.fail(
-          "Cannot remove item " + item.getProductId() + " because it is not in the cart.");
+  public Effect<Empty> removeItem(
+          ShoppingCartApi.RemoveLineItem command,
+          ShoppingCartDomain.Cart currentState,
+          EventSourcedEntityEffect.Builder<Empty, ShoppingCartDomain.Cart> effectBuilder,
+          CommandContext ctx) {
```

Comment:
EventSourcedEntityEffect.Builder + CommandContext is looking heavy. 

Not fully clear why we need it. I have to checkout the code and feel it under my fingers. 

In any case, worth experiment more to reduce it.